### PR TITLE
On title search, try some upper/lower case variants.

### LINKED ISSUE #124: ZIM backend: Title search should try some variants (at least upper/lower case)
Titles are currently stored in a case-sensitive way in a ZIM archive.
But Kiwix tries to workaround that by trying some variants. Cf Reader::searchSuggestions and Reader::searchSuggestionsSmart in https://sourceforge.net/p/kiwix/kiwix/ci/master/tree/src/common/kiwix/reader.cpp

We should try to do the same, until the OpenZIM format evolves to store normalized titles : #117

### PR DESCRIPTION
**Pull request by mossroy**
Fixes #124.

On the other hands, it really slows down the title search.
When there are not many results, it launches 3 case insensitive searches before giving the result, which takes a few seconds.

### CODE REVIEW

File: www/js/lib/util.js [hunk 1]
```diff
@@ -33,6 +33,78 @@ define(['q'], function(q) {
     }
     
     /**
+     * Returns the same String with the first letter in upper-case
+     * @param {String} string
+     * @returns {String}
+     */
+    function ucFirstLetter(string) {
+        if (string) {
```
Comment (peter-x) [inline www/js/lib/util.js]: What about
```
if (string && string.length >= 1)
  return string[0].toLocaleUpperCase() + string.slice(1);
else
  return string;
```

File: www/js/lib/zimArchive.js [hunk 2]
```diff
@@ -109,14 +109,62 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
...
      * @param {String} prefix
      * @param {Integer} resultSize
-     * @param {type} callback
-     * @returns {callbackTitleList}
+     * @param {callbackTitleList} callback
      */
     ZIMArchive.prototype.findTitlesWithPrefix = function(prefix, resultSize, callback) {
         var that = this;
+        that.findTitlesWithPrefixCaseSensitive(prefix, resultSize, function(titles) {
+            if (titles.length < resultSize) {
```
Comment (peter-x) [inline www/js/lib/zimArchive.js]: It's a matter of style, but I think it is better readable to reverse the cases:
```
if (titles.length >= resultSize) {
  callback(titles);
  return;
}
...
```
(and have to explicit else case to reduce the indentation).

File: www/js/lib/zimArchive.js [hunk 3]
```diff
@@ -109,14 +109,62 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
...
+                // Let's add results with first letter upper-case
+                var ucPrefix = util.ucFirstLetter(prefix);
```
Comment (peter-x) [inline www/js/lib/zimArchive.js]: You could check whether `prefix !== ucPrefix`

File: www/js/lib/zimArchive.js [hunk 4]
```diff
@@ -109,14 +109,62 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
...
+                that.findTitlesWithPrefixCaseSensitive(ucPrefix, resultSize, function(ucTitles) {
+                    titles.push.apply(titles, ucTitles);
+                    titles = util.removeDuplicatesInArray(titles);
```
Comment (peter-x) [inline www/js/lib/zimArchive.js]: If `prefix !== ucPrefix`, there should be no duplicates

File: www/js/lib/zimArchive.js [hunk 5]
```diff
@@ -109,14 +109,62 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
...
+                    if (titles.length < resultSize) {
+                        // Let's add results with first letter lower-case
+                        var lcPrefix = util.ucFirstLetter(prefix);
```
Comment (peter-x) [inline www/js/lib/zimArchive.js]: Typo: should be `lcFirstLetter`

File: www/js/lib/zimArchive.js [hunk 6]
```diff
@@ -109,14 +109,62 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
...
      * 
      * @param {String} prefix
      * @param {Integer} resultSize
-     * @param {type} callback
-     * @returns {callbackTitleList}
+     * @param {callbackTitleList} callback
      */
     ZIMArchive.prototype.findTitlesWithPrefix = function(prefix, resultSize, callback) {
         var that = this;
+        that.findTitlesWithPrefixCaseSensitive(prefix, resultSize, function(titles) {
```
Comment (peter-x) [inline www/js/lib/zimArchive.js]: What about the following:
```
var prefixVariants = util.removeDuplicatesInArray([prefix, util.ucFirstLetter(prefix), util.lcFirstLetter(prefix), util.ucEveryFirstLetter(prefix)]);
var titles = [];
function searchNextVariant() {
  if (prefixVariants.length == 0 || titles.length >= resultSize) { callback(titles); return; }
  var prefix = prefixVariants[0];
  prefixVariants = prefixVariants.slice(1);
  that.findTitlesWithPrefixCaseSensitive(prefix, resultSize - titles.length, function(newTitles) {
    titles.push.apply(titles, newTitles);
    searchNextVariant();    
  });
  searchNextVariant();
}
```

Comment (mossroy) [inline www/js/lib/zimArchive.js]: Yes, your suggestion is much better! Thanks a lot.
It should be a bit faster, too.

**Pull request merged**